src/uu/lsmem/src/lsmem.rs

```diff
@@ -3,15 +3,543 @@
+    ZoneDevice,
+    #[serde(rename = "ZONE_NONE")]
+    ZoneNone,
+    #[serde(rename = "ZONE_UNKNOWN")]
+    ZoneUnknown,
+    #[serde(rename = "MAX_NR_ZONES")]
+    MaxNrZones,
+}
+
+#[derive(PartialEq, Clone)]
+enum MemoryState {
+    Online,
+    Offline,
+    GoingOffline,
+    Unknown,
+}
+
+impl core::fmt::Display for MemoryState {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        match self {
+            MemoryState::Online => write!(f, "online"),
+            MemoryState::Offline => write!(f, "offline"),
+            MemoryState::GoingOffline => write!(f, "going-offline"),
+            MemoryState::Unknown => write!(f, "unknown"),
+        }
+    }
+}
+
+impl FromStr for MemoryState {
+    type Err = ();
+    fn from_str(input: &str) -> Result<MemoryState, Self::Err> {
+        match input {
+            "online" => Ok(MemoryState::Online),
+            "offline" => Ok(MemoryState::Offline),
+            "going-offline" => Ok(MemoryState::GoingOffline),
+            "unknown" => Ok(MemoryState::Unknown),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct MemoryBlock {
+    index: u64,
+    count: u64,
+    state: MemoryState,
+    node: i32,
+    nr_zones: usize,
+    zones: [ZoneId; ZoneId::MaxNrZones as usize],
+    removable: bool,
+}
+
+impl MemoryBlock {
+    fn new() -> Self {
+        MemoryBlock {
+            index: 0,
+            count: 0,
+            state: MemoryState::Unknown,
+            node: 0,
+            nr_zones: 0,
+            zones: [ZoneId::ZoneUnknown; ZoneId::MaxNrZones as usize],
+            removable: true,
+        }
+    }
+}
+
+#[derive(Tabled)]
+struct TableRow {
+    #[tabled(rename = "RANGE")]
+    range: String,
+    #[tabled(rename = "SIZE")]
+    size: String,
+    #[tabled(rename = "STATE")]
+    state: String,
+    #[tabled(rename = "REMOVABLE")]
+    removable: String,
+    #[tabled(rename = "BLOCK")]
+    block: String,
+    #[tabled(rename = "NODE")]
+    node: String,
+    #[tabled(rename = "ZONES")]
+    zones: String,
+}
+
+impl TableRow {
+    fn new() -> Self {
+        TableRow {
+            range: String::new(),
+            size: String::new(),
+            state: String::new(),
+            removable: String::new(),
+            block: String::new(),
+            node: String::new(),
+            zones: String::new(),
+        }
+    }
+}
+
+struct Lsmem {
+    ndirs: usize,
+    dirs: Vec<PathBuf>,
+    blocks: Vec<MemoryBlock>,
+    nblocks: usize,
+    block_size: u64,
+    mem_online: u64,
+    mem_offline: u64,
+
+    have_nodes: bool,
+    // raw: bool,
+    // export: bool,
+    // json: bool,
+    // noheadings: bool,
+    // summary: bool,
+    list_all: bool,
+    bytes: bool,
+    want_summary: bool,
+    want_table: bool,
+    split_by_node: bool,
+    split_by_state: bool,
+    split_by_removable: bool,
+    split_by_zones: bool,
+    have_zones: bool,
+    // table: Vec<Vec<String>>,
+}
```

Comment:
I think this struct is a candidate for splitting into at least two structs because currently it contains unrelated things: memory information on the one hand, and options provided by the user on the other hand. But it's probably something for a future PR.

Reply:
I add struct Options to store the options